### PR TITLE
Port `ArrayDistinct` to `functions-array` subcrate

### LINKED ISSUE #9545: Port `ArrayDistinct` to `function-arrays` subcrate
### Is your feature request related to a problem or challenge?

ArrayDistinct function needs to be ported to new `function-arrays` subcreate in terms of #9285.

### Describe the solution you'd like

Port `ArrayDistinct` to `function-arrays` subcrate.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by erenavsarogullari**
## Which issue does this PR close?
Closes #9545.

## What changes are included in this PR?
This PR aims to do following changes in terms of Epic https://github.com/apache/arrow-datafusion/issues/9285:
1- `ArrayDistinct` is ported to `functions-array` subcrate,
2- `general_array_distinct` is ported to `kernels.rs` but could not be removed from `array_expressions.rs` due to its legacy references: [ref](https://github.com/apache/arrow-datafusion/compare/main...erenavsarogullari:arrow-datafusion:issue-9545?expand=1#diff-ee636e21e786e49965f9c86552d090386b3174d558528e715dc98df2a93d3796R721)
3- `ScalarUDFImpl. return_type` has been implemented to support input types: `List`, `FixedSizeList` and `LargeList`,
4- Function Documentation is added.

## Are these changes tested?
Yes, all `array.slt` based `array_distinct` tests are passed.

## Are there any user-facing changes?
No

### CODE REVIEW

**Comment by Weijun-H** [comment]
Hi @erenavsarogullari , could you add a roundtrip test here?
https://github.com/apache/arrow-datafusion/blob/4cd3c433004a7a6825643d6b3911db720efe5f76/datafusion/proto/tests/cases/roundtrip_logical_plan.rs#L582-L611

**Comment by erenavsarogullari** [comment]
@Weijun-H Thanks for the review. It is added.

File: docs/source/user-guide/sql/scalar_functions.md [hunk 1]
```diff
@@ -2204,6 +2204,34 @@ array_dims(array)
 
 - list_dims
 
+### `array_distinct`
```
Comment (Weijun-H) [inline docs/source/user-guide/sql/scalar_functions.md:2207]: Could you also add links for `array_distinct` and `list_distinct` here?
https://github.com/apache/arrow-datafusion/blob/4cd3c433004a7a6825643d6b3911db720efe5f76/docs/source/user-guide/sql/scalar_functions.md#L1943-L1952

Comment (erenavsarogullari) [reply docs/source/user-guide/sql/scalar_functions.md:2209]: Addressed

**Pull request merged**